First version of Durable Function chaining lesson (#9)
marcduiker/azure-functions-university: homework/durablefunctions/chaining/durable_sapapi_ts.md

# Advanced Homework - Call SAP APIs in a Sequence

## Goal 🎯

> The goal of this homework is the development of a durable function setup to interact with an SAP backend system. 

Author: Done

Assume the following scenario:

> Out function gets triggered via a purchase order create event that gives us an identifier for the order item. We want to identify the material group texts used in the order item. To do we must call a down stream system. We cannot retrieve the text in one call but we first make a call to get the material group ID and with this result issue a second call in order to retrieve the desired texts. This represents a perfect scenario for a sequence of function calls. 

Reviewer: ```suggestion
Our function gets triggered via a purchase order create event that gives us an identifier for the order item. We want to identify the material group texts used in the order item. To do so we must call a down stream system. We cannot retrieve the text in one call but we first make a call to get the material group ID and with this result issue a second call in order to retrieve the desired texts. This represents a perfect scenario for a sequence of function calls. 

```

Author: Done + reformatted

For this example we make use of a sandbox SAP system and the OData APIs available from there. To make the calls to the system a bit more convenient we make use of the [SAP Cloud SDK](https://community.sap.com/topics/cloud-sdk) that wraps the business objects in a virtual data model (VDM) and provides a fluent API to interact with the business object.

> The exercise i. e. the Activity Functions interact with an external API. This API is free to use and mimics an [SAP system](https://api.sap.com/package/SAPS4HANACloud?section=Artifacts). The only prerequisite is to register in order to get an API key. To get the free API keys to the sandbox system you must register yourself at the [API Business Hub](https://api.sap.com/) using the Log On button.

Reviewer: ```suggestion
The exercise i.e. the Activity Functions interact with an external API. This API is free to use and mimics an [SAP system](https://api.sap.com/package/SAPS4HANACloud?section=Artifacts). The only prerequisite is to register in order to get an API key. To get the free API keys to the sandbox system you must register yourself at the [API Business Hub](https://api.sap.com/) using the Log On button.

```

Author: Done

## Assignment

> Create a HTTP-triggered Durable Function consisting of two activities that need to be executed in sequence. The input parameters are 

Reviewer: ```suggestion
Create an HTTP-triggered Durable Function consisting of two activities that need to be executed in sequence. The input parameters are:

```

Author: Done

   ```typescript
   import * as df from "durable-functions"
   import { MaterialGroupData, ProductGroupInfo } from "../utils/purchaseOrderTypes"

   const orchestrator = df.orchestrator(function* (context) {

    const materialGroup: MaterialGroupData = yield context.df.callActivity("GetPoDetails", context.bindingData.input)

    context.bindingData.input.materialGroupId = materialGroup.materialGroupId
    
    const productGroupInfo: ProductGroupInfo = yield context.df.callActivity("GetPgText", context.bindingData.input )

    return productGroupInfo

   })

   export default orchestrator
   ```

### A1.5 Test the Implementation

In this section we finally test our implementation.

#### Steps

1. Set the  value of `"esModuleInterop"` to `true` in the `tsconfig.json` file. Otherwise the build of the project will fail as some parts of the SAP Cloud SDK are not fully compliant to the default project settings i. e. the importing of modules within the files.
2. Execute the Durable orchestration and fetch the data.

   > 📝 **Tip** - You find some sample data in the `demoRequests.http` file

## A2 Retries - Dealing with Temporal Errors

> Next we implement the retry functionality. We want to separate the 

Reviewer: Sentence not complete?

Author: Rewrote sentence - done
